app/services/PaymentAllocationsService.scala

```diff
@@ -0,0 +1,80 @@
+import models.paymentAllocations.PaymentAllocations
+import play.api.Logger
+import services.PaymentAllocationsService.PaymentAllocationError
+import uk.gov.hmrc.http.HeaderCarrier
+
+import javax.inject.{Inject, Singleton}
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class PaymentAllocationsService @Inject()(incomeTaxViewChangeConnector: IncomeTaxViewChangeConnector,
+                                          val appConfig: FrontendAppConfig)
+                                         (implicit ec: ExecutionContext) {
+
+  def getPaymentAllocation(nino: String, documentNumber: String)
+                          (implicit hc: HeaderCarrier, user: MtdItUser[_]): Future[Either[PaymentAllocationError.type, PaymentAllocationViewModel]] = {
+
+    incomeTaxViewChangeConnector.getFinancialDataWithDocumentDetails(nino, documentNumber) flatMap {
+      case documentDetailsWithFinancialDetailsModel: FinancialDetailsWithDocumentDetailsModel =>
+        incomeTaxViewChangeConnector.getPaymentAllocations(documentDetailsWithFinancialDetailsModel.documentDetails.head.paymentLot.get,
+          documentDetailsWithFinancialDetailsModel.documentDetails.head.paymentLotItem.get) flatMap {
+          case paymentAllocations: PaymentAllocations => createPaymentAllocationWithClearingDate(nino, paymentAllocations) map {
+            case paymentAllocationWithClearingDate: Seq[(PaymentAllocations, Option[String])] if paymentAllocationWithClearingDate.find(_._2 == None) == None =>
+              Right(PaymentAllocationViewModel(documentDetailsWithFinancialDetailsModel, paymentAllocationWithClearingDate))
+            case _ =>
+              Logger.error("[PaymentAllocationsService][getPaymentAllocation] Could not retrieve document with financial details for payment allocations")
+              Left(PaymentAllocationError)
+          }
+          case _ =>
+            Logger.error("[PaymentAllocationsService][getPaymentAllocation] Could not retrieve payment allocations with document details")
+            Future.successful(Left(PaymentAllocationError))
+        }
+      case _ =>
+        Logger.error("[PaymentAllocationsService][getPaymentAllocation] Could not retrieve document with financial details for payment charge model")
+        Future.successful(Left(PaymentAllocationError))
+    }
+  }
+
+
+  def createPaymentAllocationWithClearingDate(nino: String, paymentCharge: PaymentAllocations)
+                                             (implicit hc: HeaderCarrier): Future[Seq[(PaymentAllocations, Option[String])]] = {
+    Future.sequence(paymentCharge.allocations map {
+      allocations =>
+        incomeTaxViewChangeConnector.getFinancialDataWithDocumentDetails(nino, allocations.transactionId.get) map {
+          case paymentAllocationChargesModel: FinancialDetailsWithDocumentDetailsModel =>
+            (paymentCharge, paymentAllocationChargesModel.paymentDetails.head.items.get.head.clearingDate)
```

Comment:
This is not correct, for each allocation it cannot just get the paymentDetails.head, it must find the paymentDetails with a matching chargeType from the allocation. It also cannot just get items.get as it must also find the item which has a matching paymentLot and paymentLotItem as the first subitem will always be the charge data and not a payment allocation.

Reply:
Please refer to the diagram attached to the story for more detail